### PR TITLE
Draft PR: add Pyright static typing to hook_points.py

### PR DESCRIPTION
**Pull request by starship006**
# Description

This draft PR adds Pyright static typing to hook_points.py. See [Alan's message on the slack](https://opensourcemechanistic.slack.com/archives/C04T79RAW8Z/p1699673111412959) for motivation/background. Specifically, this PR:
- modifies the HookPoint and HookedRootModule to support a baseline amount of pyright typing
- adds a [tool.pyright] configuration to pyproject.toml to configure the typing for this project

It's not entirely complete, but is in a state to where it would benefit from some review/assistance. Before being PR ready, we should:
- Address some TODO's I've marked on lines of code that I'm ucnertain about
- Clean up remaining type warnings on hook.py 
- If possible, adjust the pyproject.toml file so the type warnings quit applying to all files

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] My changes generate no new warnings
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### CODE REVIEW

**Comment by starship006** [comment]
I was also using this draft PR as a way to see if the 50+ tests I was failing locally was due to my machine or not; it seems like this may be the case, but I'm also failing some formatting cases. Will fix.

File: transformer_lens/hook_points.py
Comment (derpyplops) [inline transformer_lens/hook_points.py:207]: probably fine, this is also how varargs are typed in the `typing` module

File: transformer_lens/hook_points.py
Comment (alan-cooney) [inline transformer_lens/hook_points.py:143]:
```suggestion
        level: Optional[int] = None,
```

Comment (alan-cooney) [reply transformer_lens/hook_points.py:143]: nit: This is the same, but more standard. Will help us find-and-replace all optional types in the codebase once Optional and Union both get removed (in Python 3.13 I think).

File: transformer_lens/hook_points.py
Comment (alan-cooney) [inline transformer_lens/hook_points.py:177]: Assertions are just for testing (they are completely ignored by Python if optimization is used (-O)).

The better thing to do here is an if statement, and then raise an error.

File: transformer_lens/hook_points.py
Comment (alan-cooney) [inline transformer_lens/hook_points.py:169]: I think so - can't imagine a time when it wouldn't be

```suggestion
```

File: transformer_lens/hook_points.py
Comment (alan-cooney) [inline transformer_lens/hook_points.py:203]:
```suggestion
    name: Optional[str]
```

nit

File: transformer_lens/hook_points.py
Comment (alan-cooney) [inline transformer_lens/hook_points.py:207]:
```suggestion
```

File: transformer_lens/hook_points.py
Comment (alan-cooney) [inline transformer_lens/hook_points.py:238]: Could also use Literal[] here?

File: transformer_lens/hook_points.py
Comment (alan-cooney) [inline transformer_lens/hook_points.py:472]: Should use if/else and then raise a ValueError

File: transformer_lens/hook_points.py
Comment (alan-cooney) [inline transformer_lens/hook_points.py:483]: Similar assert issue

**Review by alan-cooney — CHANGES_REQUESTED** [review]
Thanks for adding this and apologies for the delay over the holiday period!

I've put some comments - mostly just a few small things. Worth doing a search for the key things raised (e.g. use of assert) as there are a few repeated places in the file.

**Comment by starship006** [comment]
Sorry for the long turnaround on this. I've addressed all of the above comments, they were very helpful, thanks. @alan-cooney is this better now?

**Comment by bryce13950** [comment]
@starship006 I just opened a PR on your fork for the most recent changes to TransformerLens. @alan-cooney has become incredibly busy in the last couple months, but I am going to poke him to approve this, so that I can get it merged when it is done. There have been quite a few changes to the `hook_points` since you originally put this together. There is at least one param that needs to be typed out, and then we will see if the CI passes.

Also, for the future, if you name your branch on your fork something different than "main", maintainers can then do things like merge the current head into your PR without needing to open a separate PR on your fork. Not a big deal, but thought I would put that info out there in case you didn't know, so that you don't need to be bothered with it in the future.

**Comment by starship006** [comment]
Thanks for taking care of the merge. I think I've taken care of everything. Two notes:
- both of the `add_caching_hooks` functions in the merge received code which contained an `is_backward` variable, but the function header wasn't updated to reflect that. ive added it back in for both instances
- linter was complaining about the code accessing`hook.name`while assuming that it isn't None, even though it is typed with `Optional`. i didn't quite know how to deal with this: I imagine you could put in `#type: ignore` on the parts where its accessing `hook.name` or just do a check to ensure it is not None. I've opted for the later, but let me know if there's a better way to do this.

**Comment by starship006** [comment]
Ah, okay, nevermind it looks like there are some type checks that need to pass. One thing that I'm noticing is the recurring theme of `NamesFilter`, which is an Optional type of either a sequence or a callable, is getting called but the linter doesn't know that it is a callable and not a sequence. Should I also go ahead and throw in isinstance checks for all of these too? (The reason I'm asking is that I don't know if doing too many of these slows down the code too much / is unprincipled and there's a better way of doing it)

**Comment by bryce13950** [comment]
Go ahead and assert is instance on that param. The hit on performance for doing something like that is incredibly small, especially in the grand scheme of things while dealing with massive files like LLM models. 

Also, I haven't been able to reach Alan, so why don't you reopen a new PR for this, and just reference this PR number in the new one. We can't close this one without him approving it, and I don't know if it is going to happen.

**Pull request closed without merging**
